### PR TITLE
.sqlproj and datasources.json file parsing

### PR DESCRIPTION
**Pull request by Benjin**
* Redoes the construction of the project tree to enable these and upcoming features
* Parses project tree from .sqlproj file
* Parses SQL connection string info from datasources.json file
* Supports multiple database projects open at the same time
* changes tree to match VS Code's preferred URI-based tree structuring

### CODE REVIEW

File: extensions/sql-database-projects/package-lock.json [hunk 1]
```diff
@@ -0,0 +1,325 @@
+{
+  "name": "sql-database-projects",
```
Comment (anthonydresser) [inline extensions/sql-database-projects/package-lock.json]: You shouldn't include this. this is the npm lock file and we use yarn.

File: extensions/sql-database-projects/src/common/utils.ts [hunk 2]
```diff
@@ -3,7 +3,40 @@
...
+	return outerParts.join('/');
+}
+
+declare global {
+	interface String {
+		trimChars(chars: string): string;
+	}
+}
+
+String.prototype.trimChars = function (chars: string): string {
```
Comment (anthonydresser) [inline extensions/sql-database-projects/src/common/utils.ts]: You shouldn't modify global prototypes.  Instead just make a helper function that does the same thing but takes the string as an input.

**Comment by udeeshagautam** [comment]
Can you please add some snips also for visuals

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	public async openProject(projectFile: vscode.Uri) {
+		console.log('Loading project: ' + projectFile.fsPath);
+
+		// Read project file
+		const newProject = new Project(projectFile.fsPath);
+		await newProject.readProjFile();
+		this.projects.push(newProject);
+
+		// Read datasources.json (if present)
+		let dataSourcesFilePath = path.join(path.dirname(projectFile.fsPath), constants.dataSourcesFileName);
```
Comment (Benjin) [inline extensions/sql-database-projects/src/controllers/projectController.ts]: const

File: extensions/sql-database-projects/src/common/utils.ts [hunk 4]
```diff
@@ -3,7 +3,40 @@
...
  *--------------------------------------------------------------------------------------------*/
 
+import * as vscode from 'vscode';
 
 export function getErrorMessage(error: Error | string): string {
 	return (error instanceof Error) ? error.message : error;
 }
+
+export function trimUri(innerUri: vscode.Uri, outerUri: vscode.Uri): string {
+	let innerParts = innerUri.path.split('/');
```
Comment (llali) [inline extensions/sql-database-projects/src/common/utils.ts]: >/ [](start = 39, length = 1)

what about \ if it's linux path?

**Comment by llali** [comment]
please add comment for public methods/properties

File: extensions/sql-database-projects/src/common/utils.ts (hunk 2 above)
Comment (llali) [reply extensions/sql-database-projects/src/common/utils.ts]: +1

---
In reply to: [369281017](https://github.com/microsoft/azuredatastudio/pull/8921#discussion_r369281017) [](ancestors = 369281017)

File: extensions/sql-database-projects/src/common/utils.ts (hunk 4 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/common/utils.ts]: Isn't Linux always `/`?  Windows is the one that uses `\` for file system paths and `/` for web paths.  Either way, I believe vscode.Uri resolves that concern for us by giving us access to both uri.path (what I'm using here, which should keep things consistent between platforms) and uri.fsPath (which follows the current platform's conventions).

**Comment by Benjin** [comment]
> Can you please add some snips also for visuals

Here's an example with multiple projects open (okay, fine - it's multiple instances of the same project; to fix), and multiple data sources being read from the json file:

image

**Comment by Benjin** [comment]
don't forget to update the telemetry iKey

File: extensions/sql-database-projects/src/models/project.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			result = await parser.parseStringPromise(projFileContents.toString());
+		}
+		catch (err) {
+			vscode.window.showErrorMessage(err);
+			return;
+		}
+
+		// find all folders and files to include
+
+		for (const itemGroup of result['ItemGroup']) {
```
Comment (anthonydresser) [inline extensions/sql-database-projects/src/models/project.ts]: are these typed? (`result` and `itemGroup`) if not you should consider adding your own typing on them. It helps when other people who aren't familiar with this code want to make changes here that they are just accessing properties that aren't actually there.

Comment (udeeshagautam) [reply extensions/sql-database-projects/src/models/project.ts]: Hey Anthony, I'll merge this one and take your comment in a separate PR and send out this week.

**Pull request merged**